[LLVM][NVPTX] Add NVPTX codegen support for clusterlaunchcontrol instruction

schwarzschild-radius opened the pull request:
This commit adds NVPTX codegen support for clusterlaunchcontrol instructions with tests under clusterlaunchcontrol.ll and clusterlaunchcontrol-multicast.ll. For more information, Please refer [PTX ISA](https://docs.nvidia.com/cuda/parallel-thread-execution/?a#parallel-synchronization-and-communication-instructions-clusterlaunchcontrol-try-cancel)

schwarzschild-radius committed:
[LLVM][NVPTX] Add NVPTX codegen support for clusterlaunchcontrol instruction

This commit adds NVPTX codegen support for clusterlaunchcontrol instructions with tests under clusterlaunchcontrol.ll and clusterlaunchcontrol-multicast.ll. For more information, Please refer [PTX ISA](https://docs.nvidia.com/cuda/parallel-thread-execution/?a#parallel-synchronization-and-communication-instructions-clusterlaunchcontrol-try-cancel)

durga4github commented on llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:
```diff
@@ -7301,3 +7301,66 @@ def INT_NVVM_ST_BULK_SHARED_CTA:
+
+  def _gen: NVPTXInst<(outs), (ins Int64Regs:$addr, Int64Regs:$mbar),
+                      "clusterlaunchcontrol.try_cancel.async.mbarrier::complete_tx::bytes" #
+                      !if(!eq(Multicast, 1), ".multicast::cluster::all", "") #
```
Can we just construct this multicast suffix once and then reuse it in all the 3 places below?

durga4github commented on llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:
```diff
@@ -7301,3 +7301,66 @@ def INT_NVVM_ST_BULK_SHARED_CTA:
+                      [(Intrinsic Int64Regs:$addr, Int64Regs:$mbar)]>, Requires<Pred>;
+
+  def _shared: NVPTXInst<(outs), (ins Int64Regs:$addr, Int64Regs:$mbar),
+                         "clusterlaunchcontrol.try_cancel.async.shared::cta.mbarrier::complete_tx::bytes" #
```
We could do the same thing for the prefix, too, but it has the `shared::cta` as an extra modifier. So, I would leave it to your choice.

durga4github commented on llvm/lib/Target/NVPTX/NVPTXSubtarget.h:
```diff
@@ -112,6 +112,20 @@ class NVPTXSubtarget : public NVPTXGenSubtargetInfo {
     return HasTcgen05 && PTXVersion >= 86;
   }
 
+  bool hasBlackwellArch() const {
```
Can we name it something like: isBlackwellArchAccelerated or something similar ?

durga4github commented on llvm/lib/Target/NVPTX/NVPTXSubtarget.h:
```diff
@@ -112,6 +112,20 @@ class NVPTXSubtarget : public NVPTXGenSubtargetInfo {
   }
 
+  bool hasBlackwellArch() const {
+    auto HasSM = [&]() {
```
HasRequiredSM might be a better name for this boolean.

durga4github commented on llvm/include/llvm/IR/IntrinsicsNVVM.td:
```diff
@@ -5381,4 +5381,50 @@ def int_nvvm_st_bulk_shared_cta : DefaultAttrsIntrinsic<[],
+
+def int_nvvm_clusterlaunchcontrol_try_cancel_async
+    : Intrinsic<[], [llvm_ptr_ty, llvm_ptr_ty],
+                [IntrArgMemOnly, NoCapture<ArgIndex<0>>, NoCapture<ArgIndex<1>>],
```
I am wondering whether we should explicitly attach `HasSideEffects` to these intrinsics.

gonzalobg commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -1478,6 +1478,104 @@ similar but the latter uses generic addressing (see `Generic Addressing <https:/
+"""""""""
+
+The ``clusterlaunchcontrol.try_cancel`` intrinsics requests atomically cancelling
+the launch of a cluster that has not started running yet. It asynchronously writes
```
```suggestion
the launch of a cluster that has not started running yet. It asynchronously non-atomically writes
```

schwarzschild-radius replied:
Updated the doc with the suggested change

gonzalobg commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -1478,6 +1478,104 @@ similar but the latter uses generic addressing (see `Generic Addressing <https:/
+Overview:
+"""""""""
+
+The ``llvm.nvvm.clusterlaunchcontrol.query_cancel.is_canceled`` intrinsic can be
+used to decode the opaque response written by the
```
```suggestion
The ``llvm.nvvm.clusterlaunchcontrol.query_cancel.is_canceled`` intrinsic
decodes the opaque response written by the
```

schwarzschild-radius replied:
Updated the doc with the suggested change

gonzalobg commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -1478,6 +1478,104 @@ similar but the latter uses generic addressing (see `Generic Addressing <https:/
+used to decode the opaque response written by the
+``llvm.nvvm.clusterlaunchcontrol.try_cancel`` operation.
+
+The intrinsic returns false if the request failed. If the request succeeded,
```
```suggestion
The intrinsic returns ``0`` (false) if the request failed. If the request succeeded,
```

gonzalobg commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -1478,6 +1478,104 @@ similar but the latter uses generic addressing (see `Generic Addressing <https:/
+The ``clusterlaunchcontrol.try_cancel`` intrinsics requests atomically cancelling
+the launch of a cluster that has not started running yet. It asynchronously writes
+a 16-byte opaque response to shared memory, pointed to by ``addr`` indicating whether the
+operation succeeded or failed. ``addr`` and ``mbar`` must be in ``shared::cta``
```
```suggestion
operation succeeded or failed. ``addr`` and ``mbar`` must be in ``shared::cta``,
```

schwarzschild-radius replied:
Updated the doc with the suggested change

gonzalobg commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -1478,6 +1478,104 @@ similar but the latter uses generic addressing (see `Generic Addressing <https:/
+the launch of a cluster that has not started running yet. It asynchronously writes
+a 16-byte opaque response to shared memory, pointed to by ``addr`` indicating whether the
+operation succeeded or failed. ``addr`` and ``mbar`` must be in ``shared::cta``
+otherwise the result is undefined. The completion of the asynchronous operation
```
We don't guarantee that a result will be written at all:

```suggestion
otherwise the behavior is undefined. The completion of the asynchronous operation
```

schwarzschild-radius replied:
Updated the doc with the suggested change

gonzalobg commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -1478,6 +1478,104 @@ similar but the latter uses generic addressing (see `Generic Addressing <https:/
+from other ``clusterlaunchcontrol.try_cancel`` operations from the same grid will
+contain that id.
+
+The ``multicast`` variant specifies that the response is asynchronously written to
```
```suggestion
The ``multicast`` variant specifies that the response is asynchronously non-atomically written to
```

gonzalobg commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -1478,6 +1478,104 @@ similar but the latter uses generic addressing (see `Generic Addressing <https:/
+``llvm.nvvm.clusterlaunchcontrol.try_cancel`` operation.
+
+The intrinsic returns false if the request failed. If the request succeeded,
+it returns true. A true result indicates that:
```
```suggestion
it returns ``1`` (true). A true result indicates that:
```

gonzalobg commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -1478,6 +1478,104 @@ similar but the latter uses generic addressing (see `Generic Addressing <https:/
+
+- the thread block cluster whose first CTA id matches that of the response
+  handle will not run
+- no other successful response of another ``try_cancel`` request will contain
```
```suggestion
- no other successful response of another ``try_cancel`` request in the grid will contain
```

schwarzschild-radius replied:
Updated the doc with the suggested change

gonzalobg commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -1478,6 +1478,104 @@ similar but the latter uses generic addressing (see `Generic Addressing <https:/
+used to decode the opaque response written by the
+``llvm.nvvm.clusterlaunchcontrol.try_cancel`` operation.
+
+If the request succeeded,
+``llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid.{x,y,z}`` returns
+the coordinate of the first CTA in the canceled cluster, either x, y, or z.
+``llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid`` returns a vector
+where the first three components are the x, y, z coordinates of the first CTA.
+The contents of the fourth element are unspecified
```
```suggestion
If the request succeeded,

    - ``llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid.{x,y,z}`` returns
      the coordinate of the first CTA in the canceled cluster, either x, y, or z.
    - ``llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid`` returns a vector
      where the first three components are the x, y, z coordinates of the first CTA.

The contents of the fourth element are unspecified.
```

schwarzschild-radius replied:
Updated the doc with the suggested change

gonzalobg commented on llvm/include/llvm/IR/IntrinsicsNVVM.td:
```diff
@@ -5381,4 +5381,50 @@ def int_nvvm_st_bulk_shared_cta : DefaultAttrsIntrinsic<[],
+def int_nvvm_clusterlaunchcontrol_try_cancel_async
+    : Intrinsic<[], [llvm_ptr_ty, llvm_ptr_ty],
+                [IntrHasSideEffects, IntrArgMemOnly, NoCapture<ArgIndex<0>>, NoCapture<ArgIndex<1>>],
+                "llvm.nvvm.clusterlaunchcontrol.try_cancel.async">;
```
Where can we specify that the shared memory is 16 byte aligned and the mbarrier is 8 bytes aligned?

schwarzschild-radius replied:
> the ptr addresses of the data and barrier are 16 byte aligned,
We can enforce that in the IR Verifier that the ptr addresses should be 16-byte aligned

> the generic pointers of data and barrier fall in the shared::cta memory window
Since it is runtime value, I am not sure how we can enforce in from the compiler side

CC += @Artem-B / @durga4github for comments

gonzalobg replied:
In general, for all intrinsics that only accept "addresses to one state space and generic addresses to that one state space", LLVM should probably be doing what you propose ( @akshayrdeodhar for vis, since it sounds like something worth exploring related to better optimizations for intrinsics ).

Some intrinsics accepting generic addresses offload the address space detection + conversion to HW. 
If SW were to perform the detection + conversion, then that is pretty much always worse than doing it in HW.
For these, it may be good if we could express that a generic address, e.g., 
- can only narrow to some state spaces (e.g. either "global" or "shared").
- cannot narrow to some state spaces (e.g. "local").

Such that if address space narrowing proves that, e.g., a generic address does not point to e.g., "shared", then an intrinsic that only supports shared | global is called, this intrinsic further enables address space narrowing to narrow it further to global.

gonzalobg commented on llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:
```diff
@@ -7301,3 +7301,67 @@ def INT_NVVM_ST_BULK_SHARED_CTA:
+            !strconcat("{{\n\t",
+            !strconcat(".reg .b128 %handle;\n\t",
+            !strconcat("mov.b128 %handle, {$try_cancel_response0, $try_cancel_response1};\n\t",
+            !strconcat("clusterlaunchcontrol.query_cancel.get_first_ctaid.v4.b32.b128 {$r1, $r2, $r3, $r4}, %handle;\n\t", "}}")))), []>,
```
Can the 4th register be bottom (`_`)?

```suggestion
            !strconcat("clusterlaunchcontrol.query_cancel.get_first_ctaid.v4.b32.b128 {$r1, $r2, $r3, _}, %handle;\n\t", "}}")))), []>,
```

schwarzschild-radius replied:
Removed the 4th argument from the intrinsic and return a struct of 3 elements instead. Also, added placeholder in the PTX string

schwarzschild-radius replied:
Removed the vector variant in the favour of single element variants

gonzalobg commented on llvm/test/CodeGen/NVPTX/clusterlaunchcontrol.ll:
```diff
@@ -0,0 +1,140 @@
+; CHECK-NEXT:    }
+; CHECK-NEXT:    st.param.v4.b32 [func_retval0], {%r1, %r2, %r3, %r4};
+; CHECK-NEXT:    ret;
+  %v0 = call <4 x i32> @llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid(i128 %try_cancel_response)
```
I think this should return `undef` for the last element of the vector, e.g., allowing `st.param.v4.b32 ..., {%r1, %r2, %r3, _}` to be generated (or just reusing one of the other registers, doesn't really matter which one).

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:
```diff
@@ -7301,3 +7301,67 @@ def INT_NVVM_ST_BULK_SHARED_CTA:
+
+def CLUSTERLAUNCHCONTROL_QUERY_CANCEL_IS_CANCELED :
+  NVPTXInst<(outs Int1Regs:$pred), (ins Int64Regs:$try_cancel_response0, Int64Regs:$try_cancel_response1),
+            !strconcat("{{\n\t",
```
could you use `#` here instead of nesting `!strconcat`?

schwarzschild-radius replied:
Replaced strconcat with `#`

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -6036,6 +6102,60 @@ static void ReplaceINTRINSIC_W_CHAIN(SDNode *N, SelectionDAG &DAG,
+
+    SDVTList ResVTs = DAG.getVTList(ListVTs);
+
+    SmallVector<SDValue, 8> Ops;
```
A `SmallVector` seems overkill given that there will always be exactly 4 operands. Can we instead use an initializer list directly in the `getMemIntrinsicNode` call?

Artem-B commented on llvm/include/llvm/IR/IntrinsicsNVVM.td:
```diff
@@ -5381,4 +5381,50 @@ def int_nvvm_st_bulk_shared_cta : DefaultAttrsIntrinsic<[],
+
+def int_nvvm_clusterlaunchcontrol_try_cancel_async
+    : Intrinsic<[], [llvm_ptr_ty, llvm_ptr_ty],
+                [IntrHasSideEffects, IntrArgMemOnly, NoCapture<ArgIndex<0>>, NoCapture<ArgIndex<1>>],
```
Instruction documentation says "asynchronously writes an opaque response to shared memory". The way I read it, it means that intrinsic does effectively capture the pointer. I.e. it will continue to be in use after the instruction has returned.

gonzalobg replied:
Yes, after the instruction returns, some time later, there will be a write to the shared memory pointer, and an arrival to the barrier (both).

schwarzschild-radius replied:
Removed `NoCapture` property as instruction captures both addr and mbar operand and writes to it later

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXSubtarget.h:
```diff
@@ -112,6 +112,20 @@ class NVPTXSubtarget : public NVPTXGenSubtargetInfo {
+      return false;
+    };
+
+    return HasRequiredSM() && PTXVersion >= 86;
```
I'm not sure we need this custom predicate at the moment.
We already use `[hasSM<100>, hasArchAccelFeatures, hasPTX<86>]` in a lot of places.
We could shorten that in tablegen, if necessary, if it becomes too much of a burden, but it does not seem to be the case for now.

If/when some instructions eventually turn out to be blackwell-only, we'll sort it out *then*, when we have specific info.

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -1017,6 +1017,9 @@ NVPTXTargetLowering::NVPTXTargetLowering(const NVPTXTargetMachine &TM,
                      Custom);
+
+  // Enable custom lowering for i128 bit type supported in PTX
+  setOperationAction(ISD::INTRINSIC_W_CHAIN, MVT::i128, Custom);
```
Nevermind, disregard this.

schwarzschild-radius commented:
Summary of changes in the latest revision:
- Addressed documentation changes by @gonzalobg
- Removed generic variants of the intrinsics
- Removed the custom selection code for `clusterlaunchcontrol_query_cancel*` variants in the favour of custom SelectionDAG Nodes as requested by @AlexMaclean 
- Replaced the vector variant of `clusterlaunchcontrol_query_cancel_get_first_ctaid` with a struct variants taking three arguments
- Replaced the new sm arch predicate with the existing one as requested by @Artem-B
- Removed `NoCapture` properties from `clusterlaunchcontrol.try_cancel` variants they do capture the address and mbarrier arguments

schwarzschild-radius commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -312,6 +312,37 @@ void NVPTXDAGToDAGISel::SelectTcgen05Ld(SDNode *N, bool hasOffset) {
   }
 }
 
+void NVPTXDAGToDAGISel::SelectClusterLaunchControl(SDNode *N) {
```
Replaced Custom C++ Selection with Tablegen based one by creating new SDAG Nodes

schwarzschild-radius commented on llvm/include/llvm/IR/IntrinsicsNVVM.td:
```diff
@@ -5381,4 +5381,50 @@ def int_nvvm_st_bulk_shared_cta : DefaultAttrsIntrinsic<[],
+                "llvm.nvvm.clusterlaunchcontrol.try_cancel.async">;
+
+def int_nvvm_clusterlaunchcontrol_try_cancel_async_shared
+    : Intrinsic<[], [llvm_shared_ptr_ty, llvm_shared_ptr_ty],
```
Updated all of the Intrinsics to be DefaultAttrIntrinsics in the latest revision

schwarzschild-radius commented on llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:
```diff
@@ -7301,3 +7301,67 @@ def INT_NVVM_ST_BULK_SHARED_CTA:
+  def _shared32: NVPTXInst<(outs), (ins Int32Regs:$addr, Int32Regs:$mbar),
+                           "clusterlaunchcontrol.try_cancel.async.shared::cta.mbarrier::complete_tx::bytes" #
+                           !if(!eq(Multicast, 1), MulticastMod, "") #
+                           ".b128 [$addr], [$mbar];",
```
Thanks! Simplified the patterns with addr/ADDR operands for shared memory pointers

AlexMaclean commented on llvm/include/llvm/IR/IntrinsicsNVVM.td:
```diff
@@ -5472,4 +5472,38 @@ def int_nvvm_st_bulk_shared_cta : DefaultAttrsIntrinsic<[],
+
+// clusterlaunchcontrol.query_cancel.get_first_ctaid*
+
+def int_nvvm_clusterlaunchcontrol_query_cancel_get_first_ctaid
+    : DefaultAttrsIntrinsic<[llvm_i32_ty, llvm_i32_ty, llvm_i32_ty], [llvm_i128_ty], [],
+                            "llvm.nvvm.clusterlaunchcontrol.query_cancel.get_first_ctaid">;
+
+foreach dim = ["x", "y", "z"] in {
+def int_nvvm_clusterlaunchcontrol_query_cancel_get_first_ctaid_ # dim
```
It seems like these 2 sets of intrinsics offer 2 methods of querying the exact same information. Can we pick one as a canonical form and use that everywhere? I'd go with the single element variants. This would be consistent with how we support special registers which PTX allows us to treat as a vector, but we never do.

schwarzschild-radius replied:
@AlexMaclean Removed the vector/struct variant of the intrinsic in the favour of single element variants

AlexMaclean commented on llvm/include/llvm/IR/IntrinsicsNVVM.td:
```diff
@@ -5472,4 +5472,38 @@ def int_nvvm_st_bulk_shared_cta : DefaultAttrsIntrinsic<[],
+
+foreach dim = ["x", "y", "z"] in {
+def int_nvvm_clusterlaunchcontrol_query_cancel_get_first_ctaid_ # dim
+    : DefaultAttrsIntrinsic<[llvm_i32_ty], [llvm_i128_ty], [],
```
My interpretation of the spec is that this instruction is just pulling some bits from the input. If so can we mark it as IntrNoMem and speculatable?

schwarzschild-radius replied:
Added `IntroNoMem` and `IntrSpeculatable` properties and updated the lowering accordingly

Artem-B commented:
@schwarzschild-radius would it be possible to go over still open comments and mark the ones you've addressed as resolved? Right now it's really hard to tell whether you're still working on all/some of them, or if they are done.

schwarzschild-radius commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -1478,6 +1478,104 @@ similar but the latter uses generic addressing (see `Generic Addressing <https:/
+  declare void @llvm.nvvm.clusterlaunchcontrol.try_cancel.async(ptr %addr, ptr %mbar)
+  declare void @llvm.nvvm.clusterlaunchcontrol.try_cancel.async.shared(ptr addrspace(3) %addr, ptr addrspace(3) %mbar)
+  declare void @llvm.nvvm.clusterlaunchcontrol.try_cancel.async.multicast(ptr %addr, ptr %mbar)
+  declare void @llvm.nvvm.clusterlaunchcontrol.try_cancel.async.multicast.shared(ptr addrspace(3) %addr, ptr addrspace(3) %mbar)
```
That's right. We have used `shared` in the past to refer to `shared::cta` this keeps the name concise. If the instruction extends support for shared::cluster, we can add shared_cluster variants which would disambiguate. Also, NVPTX now support addrspace for shared::cluster, so the operand would disambiguate as well

schwarzschild-radius commented on llvm/docs/NVPTXUsage.rst:
```diff
@@ -1478,6 +1478,104 @@ similar but the latter uses generic addressing (see `Generic Addressing <https:/
+it returns true. A true result indicates that:
+
+- the thread block cluster whose first CTA id matches that of the response
+  handle will not run
```
Updated the doc with the suggested change

schwarzschild-radius commented:
@Artem-B I have addressed all of the comments and resolved them as well. Please let me know if the review is good to go

AlexMaclean reviewed (approved):
LGTM. Please wait for other reviewers before landing.

As a side note, all the NVPTXISelLowering code and custom nodes seem like they could be eliminated if only we fully supported `i128` types as legal. This has been on my list for a while and perhaps this is a sign I need to come back to it.

durga4github commented on llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:
```diff
@@ -7411,3 +7411,59 @@ def INT_NVVM_ST_BULK_SHARED_CTA:
+class CLUSTERLAUNCHCONTROL_QUERY_CANCEL_GET_FIRST_CTAID<string Dim>:
+  NVPTXInst<(outs Int32Regs:$reg), (ins Int64Regs:$try_cancel_response0, Int64Regs:$try_cancel_response1),
+            "{{\n\t" #
+               ".reg .b128 %handle;\n\t" #
```
nit: Should we say `%clc_handle` ?

schwarzschild-radius replied:
Updated the name to `%clc_handle` but this shouldn't matter as the variable is scoped and shouldn't interfere with other `%handle` variables

durga4github replied:
Thanks, I meant to add more context to `handle`..

durga4github commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -2795,12 +2801,56 @@ static SDValue LowerIntrinsicVoid(SDValue Op, SelectionDAG &DAG) {
+  case Intrinsic::nvvm_clusterlaunchcontrol_query_cancel_get_first_ctaid_x:
+  case Intrinsic::nvvm_clusterlaunchcontrol_query_cancel_get_first_ctaid_y:
+  case Intrinsic::nvvm_clusterlaunchcontrol_query_cancel_get_first_ctaid_z:
+    return LowerClusterLaunchControl(Op, DAG);
```
LowerClusterLaunchControlQuery(Op, DAG) ?

schwarzschild-radius replied:
Renamed it to `LowerClusterLaunchControlQueryCancel`, I think we will not need it once we have native i128 support in NVPTX

durga4github commented on llvm/test/CodeGen/NVPTX/clusterlaunchcontrol.ll:
```diff
@@ -0,0 +1,111 @@
+; RUN: %if ptxas-12.8 %{ llc < %s -mtriple=nvptx64 -mcpu=sm_100 -mattr=+ptx86 | %ptxas-verify -arch=sm_100 %}
+; RUN: %if ptxas-12.8 %{ llc < %s -mtriple=nvptx64 -mcpu=sm_100 -mattr=+ptx86 --nvptx-short-ptr | %ptxas-verify -arch=sm_100 %}
+
+define void @nvvm_clusterlaunchcontrol_try_cancel(ptr %addr, ptr %mbar,
```
are we using these two args? `addr` and `mbar`?

schwarzschild-radius replied:
Thanks for catching that, we not longer need the generic pointers. Fixed in the latest revision

durga4github replied:
Thanks!

durga4github reviewed (approved):
The latest changes LGTM

schwarzschild-radius commented:
@Artem-B I have addressed all of the comments, Please let me know if I am can go ahead with merge
